Update README.md (#11)
pgcentralfoundation/postgrestd: README.md

# The Postgres Standard Library

> `Postgrestd` is a custom variant to the Rust Standard Library. This is the main source code repository for `postgrestd`. 

Reviewer: ```suggestion
`postgrestd` is a fork of the Rust standard library that implements a variant on the internal-only `std::sys` module plus additional targets that use this module. It replaces many of the bindings to the operating system with bindings to PostgreSQL's interfaces or to no-op functions, restricting what Rust code compiled with it can access through Safe Rust.
```

## Key Features

> The goal of the library is to prevent code from accessing 

Reviewer: ```suggestion
The goal of this fork is to prevent Safe Rust code from accessing 
```

  * file handle operations
  * internals of the postgres process 

>   * the OS with the permissions of the server process

Reviewer: What you are really citing here is preventing access to the operating system's ability to start subprocesses or to access the shell environment that Postgres has been started in, either of which can amount to allowing code to roam free in the computing environment.

Author: Updated with comments

> The library exposes certain calls into the environment, in the pattern of libc, instead of directly interfacing with the C standard library, the `postgrestd` equivalent functions are called. When it is not applicable we deny the code access, avoiding errors. Code that relies on having access to the environment will panic and abort the transaction.  

Reviewer: ```suggestion
In the typical implementation of the Rust standard library, it uses a module, `std::sys`, that defines various system-specific bindings which either address platform-specific datatype compatibility concerns or directly bind against the host's system library (which is usually the C standard library, AKA libc). Instead of directly interfacing with the C standard library, `postgrestd::sys` calls functions from the Postgres C API instead that implement similar functionality. When the functionality is not applicable or not desired, `Result::Err` that indicate unsupported functionality is returned, allowing Rust code that handles it appropriately to continue functioning. Code that relies on having access to the environment will panic, which is converted into raising an error that aborts the transaction.
```

> The key to managing this is replacing the allocator so that all memory allocations are inside the Postgres context, ensuring that data is torn down properly.

Reviewer: ```suggestion
Aborting a Rust runtime may be done at any point without being considered an unsafe operation, even if this means Drop implementations are not run. However, by replacing the Rust global allocator with one that uses Postgres memory contexts, aborted transactions do not actually leak memory: Postgres does the teardown of the aborted transaction's memory contexts, in a similar way that an operating system reaps a terminated process and reclaims its resources. It is possible for non-memory resources to leak, and handling those is not currently addressed, except by making it impossible to claim irrelevant resources such as e.g. file handles or threads.
```

## Get Started

> 

Reviewer 2: I know this is blank for a reason but I am starting to wonder exactly _what_ should go in here? Setup instructions for development? Examples of how it's used?

Author: It depends on the focus of the readme.  I am guessing that if there are users of this library they are developers looking to build plrust based things for themselves.  I was thinking this was the here is how you get this setup to build, but ... it could alos have the same focus as pgx (the extension writer) rather than the postgrestd developer.

Also, I had trouble getting it to build so I need some help there. (Thinking that its my OS/Arch)

Reviewer: ```suggestion

`postgrestd` is intended to be used as part of [PL/Rust](https://github.com/tcdi/plrust).
```

## Getting Help

> Join the [Postgres Rust Extension][pgx] (pgx) community on [Discord](https://discord.gg/kwsy38x5Kh).

Reviewer: ```suggestion
This is a part of the PL/Rust project. Join the [pgx] (Postgres Extensions in Rust) community on [Discord](https://discord.gg/kwsy38x5Kh) and ask for help in [plrust-#general](https://discord.com/channels/561648697805504526/835595007791726704)!
```

## Contributing

In addition to [TCDI](https://www.tcdi.com), the project is maintained by the [pgx][pgx] community. 

We are most definitely open to contributions of any kind. Bug Reports, Feature Requests, Documentation, and even sponsorships.

> If you'd like to contribute code via a Pull Request, please make it against our main branch.

Reviewer: Currently there's no actually consistent branch, unfortunately.
```suggestion
```

> If you'd like to contribute code via a Pull Request, please make it against our develop branch. The master branch is meant to represent what is currently released.

Reviewer 2: In this context, which repo with the `develop` branch are you referring to? As far as I know, the `postgrestd` repo does not have a `develop` branch.

I know that (at least for now) the `pgx` repo uses a `develop` branch, but I think other repos do not.

Author: Yes, looks like that was a copy paste error.  This repo does need a master branch at least (?) so people know the current source of the project.

## License

> `postgrestd` is primarily distributed under the terms of both the MIT license and the Apache License (Version 2.0), with portions covered by various BSD-like licenses.

Reviewer: ```suggestion
`postgrestd` is distributed under the same terms that apply to the Rust standard library. This means it is primarily distributed under the terms of both the MIT license and the Apache License (Version 2.0), with portions covered by various BSD-like licenses. By contributing to this repository you accept that contributions to this repository may be upstreamed to the [rustc repo](https://github.com/rust-lang/rust) to be redistributed with the Rust standard library.
```
